### ISSUE TITLE
Unbound variables introduced by syntax-rules

### ISSUE THREAD

**Issue by yorickhardy** [justinethier/cyclone#361]
Binding variables in let-syntax and define-syntax creates unbound variables.
I suspect this is not intended. For example,

> $ cat unbound.scm
```
 (import (scheme base)
         (scheme write) )
 
 (let-syntax 
  ( (indirect-def
     (syntax-rules ()
      ( (indirect-def x y)
        (begin
         (define z y)
         (define x z) ) ) ) ) )
  (indirect-def a 3)
  (display a)
  (newline) )
```

outputs 3 with chibi-scheme and gauche

> $ chibi-scheme unbound.scm
> 3

> $ gosh unbound.scm
> 3

but causes an error with icyc and cyclone

> $ icyc -s unbound.scm
> Error: Unbound variable: z
> Call history:
> scheme/cyclone/util.sld:formals->list
> scheme/cyclone/util.sld:pack-lambda-arguments
> scheme/cyclone/util.sld:env:extend-environment
> scheme/cyclone/util.sld:env:define-variable!
> scheme/cyclone/util.sld:env:add-binding-to-frame!
> scheme/cyclone/util.sld:env:lookup-variable-value
> scheme/cyclone/util.sld:env:_lookup-variable-value
> scheme/base.sld:error
> scheme/base.sld:raise

> $ cyclone unbound.scm
> Error: Unbound variable(s): (z)
> Call history:
> scheme/cyclone/transforms.sld:remove
> scheme/cyclone/transforms.sld:difference
> scheme/cyclone/transforms.sld:remove
> scheme/cyclone/transforms.sld:difference
> scheme/cyclone/transforms.sld:remove
> scheme/cyclone/transforms.sld:difference
> scheme/base.sld:list
> scheme/base.sld:error
> scheme/base.sld:raise

The behaviour is a bit different with define-syntax.

> $ cat syntax.scm
```
 (import (scheme base)
         (scheme write) )

 (define-syntax indirect-def
  (syntax-rules ()
   ( (indirect-def x y)
     (begin
      (define z y)
      (define x z) ) ) ) )

 (indirect-def a 3)
 (display a)
 (newline)
```

icyc outputs 3,
> $ icyc -s syntax.scm
> 3
> cyclone> 

but cyclone indicates an error,
> $ cyclone syntax.scm
> Error: Unbound variable(s): (z)
> Call history:
> scheme/cyclone/transforms.sld:remove
> scheme/cyclone/transforms.sld:difference
> scheme/cyclone/transforms.sld:remove
> scheme/cyclone/transforms.sld:difference
> scheme/cyclone/transforms.sld:remove
> scheme/cyclone/transforms.sld:difference
> scheme/base.sld:list
> scheme/base.sld:error
> scheme/base.sld:raise

Is this behaviour correct?

**Comment by justinethier** [comment]
This will be fixed in the next release of Cyclone (0.16).

**Comment by yorickhardy** [comment]
Thank you very much!

-- 
Kind regards,

Yorick Hardy
